### ISSUE TITLE
custom canvas: keep ui layout

### ISSUE THREAD

**Issue by Esshahn** [nurpax/petmate#138]
With the new custom canvas feature, it is quite hard to work with more extreme proportions (e.g. for DirArt), like this

![screenshot 2019-02-28 um 09 43 21](https://user-images.githubusercontent.com/434355/53553178-5238aa80-3b3d-11e9-9380-832b4a2415f2.jpg)

To make things worse, when zooming in, the canvas size itself stays the same, so one can see even less

![screenshot 2019-02-28 um 09 44 18](https://user-images.githubusercontent.com/434355/53553225-74322d00-3b3d-11e9-90c7-de6a3478f79c.jpg)

Recommendation:
* keep the basic layout and proportions you already have
* palette stays on the right
* canvas is centered within the original area
* when zooming, canvas can extend within that viewport

![screenshot 2019-02-28 um 09 45 45](https://user-images.githubusercontent.com/434355/53553425-f15da200-3b3d-11e9-961b-55cfee8ec2d4.jpg)

By doing this, you can still have the overview of the full e.g. 16x32 canvas when fully zoomed out (which should be the default), but you can zoom in and still see a lot of the canvas, not being restricted by it's char size.

**Comment by nurpax** [comment]
I kind of like the explicit control over "pixel size" suggested in https://github.com/nurpax/petmate/issues/122

E.g., have something like ctrl-0 mean fit the whole petscii screen into the canvas region.  ctrl+1 means 100% pixels, ctrl+2 200% etc.  Hitting these would resize the "canvas container".

Having the canvas resize based on zoom is problematic at least from layout implementation POV.  Right now zoom doesn't change HTML/CSS layout so it's super fast.  But if the canvas window/div resizes, it causes full HTML re-layout/re-paint and that will suck for performance as everything will be re-rendered.  So I'd prefer the explicit buttons for just setting the canvas zoom size.

Sorry if my terminology is unclear here.

Basically, there's two kinds of zooms:

1) the pan/zoom transformation
2) size of the actual painting canvas

Hitting ctrl+0 would reset it to the configuration that I have currently implemented (e.g., canvas completely encloses the petscii frame).

**Comment by nurpax** [comment]
OK so I misread Ingo's above comment.  The suggestion is to keep the canvas layout always the same, e.g., it will fill the same area as now regardless of screensize.

I'm still trying to figure out what'd work best for that.  I agree that that probably would work well, but it needs a little design thinking to make it easy to understand for someone who's never worked with Petmate and these weird aspect ratio screens.

Here's a little sketch of what I mean by the "aspect ratio problem":

![image](https://user-images.githubusercontent.com/297823/53687999-2ee04c00-3d45-11e9-8dbe-9a11024acd1a.png)

**Comment by nurpax** [comment]
I think the version in current master is pretty good now.  Not entirely happy with this, but I also would like to move on to something else.  My canvas scaling business is pretty hairy and this it is starting to feel like "real work" to untangle all of that. :)

As for what's left:

- when panning, some indication of what's the current scroll position would be great.

**Comment by nurpax** [comment]
https://github.com/nurpax/petmate/issues/142 covers most of this.  please file a new one if I missed something.
